src/test/java/com/atlassian/db/replica/api/TestDualConnection.java

```diff
@@ -603,58 +642,88 @@ public void shouldHoldOnlyOneConnection() throws SQLException {
+            any(),
+            eq(new RouteDecisionBuilder(Reason.READ_OPERATION).sql(SIMPLE_QUERY).build())
+        );
     }
 
     @Test
     public void shouldExecuteOnMainWhenNotConsistent() throws SQLException {
         final ConnectionProviderMock connectionProvider = new ConnectionProviderMock();
-        final DualCall dualCall = mock(DualCall.class);
+        final DatabaseCall databaseCall = mock(DatabaseCall.class);
         final ReplicaConsistency consistency = permanentInconsistency().build();
         final Connection connection = DualConnection
             .builder(connectionProvider, consistency)
-            .dualCall(dualCall)
+            .databaseCall(databaseCall)
             .build();
 
         connection.prepareStatement(SIMPLE_QUERY).executeQuery();
 
         assertThat(connectionProvider.getProvidedConnectionTypes())
             .containsOnly(REPLICA, MAIN);
-        verify(dualCall).callMain(any());
-        verify(dualCall, never()).callReplica(any());
+        verify(databaseCall).call(
+            any(),
+            eq(new RouteDecisionBuilder(Reason.REPLICA_INCONSISTENT).sql(SIMPLE_QUERY).build())
+        );
     }
 
     @Test
-    public void shouldExecuteOnMain() throws SQLException {
+    public void shouldExecuteUpdateOnMain() throws SQLException {
         final ConnectionProviderMock connectionProvider = new ConnectionProviderMock();
-        final DualCall dualCall = mock(DualCall.class);
-        when(dualCall.callMain(any())).thenReturn(1);
+        final DatabaseCall databaseCall = mock(DatabaseCall.class);
+        when(databaseCall.call(any(), any())).thenReturn(1);
         final Connection connection = DualConnection
             .builder(connectionProvider, permanentConsistency().build())
-            .dualCall(dualCall)
+            .databaseCall(databaseCall)
             .build();
 
         connection.prepareStatement(SIMPLE_QUERY).executeUpdate();
 
         assertThat(connectionProvider.getProvidedConnectionTypes())
             .containsOnly(MAIN);
-        verify(dualCall, never()).callReplica(any());
-        verify(dualCall).callMain(any());
+        verify(databaseCall).call(any(), eq(new RouteDecisionBuilder(Reason.API_CALL).sql(SIMPLE_QUERY).build()));
     }
 
+    @Test
+    public void shouldReuseMainConnection() throws SQLException {
+        final ConnectionProviderMock connectionProvider = new ConnectionProviderMock();
+        final DatabaseCall databaseCall = mock(DatabaseCall.class);
+        when(databaseCall.call(any(), any())).thenReturn(1);
+        final Connection dualConnection = DualConnection.builder(
+            connectionProvider,
+            permanentConsistency().build()
+        ).databaseCall(databaseCall)
+            .build();
+        dualConnection.prepareStatement(SIMPLE_QUERY).executeUpdate();
+        Mockito.reset(databaseCall);
+
+        dualConnection.prepareStatement(SIMPLE_QUERY).executeQuery();
+
+        verify(databaseCall).call(
```

Comment:
@dagguh Look at this!

Reply:
Oooooh ✨ 
So it's like a reason chain just like exception cause chain. Very nice!